Regex saved fields to optionally reduce TecPlot output file size

simbilod opened the pull request:
Partially addresses #94 

Adds an `include` string argument to `write_devices` that is used to filter node and edge models when saving to TecPlot. `include `is a regular expression, and only node/edge model names matching the string will be saved. It defaults to `".*"` in order to recover legacy behaviour (save everything).

Limitations/possible improvements:
* The `include` argument  is currently not used for other file formats, but the same regex parsing could be added there too.
* The `include` argument is not used to filter out "global" data (x,y, etc.), but could be added there too.

### Example use and output, for the same device:

Command:

`ds.write_devices(file=file, device=device_name, type=type)`

Output file header:

```
TITLE = "450991316286470493"
VARIABLES = "x", "y", "Acceptors", "AtContactNode", "ContactSurfaceArea", "DEG", "Donors", "EC", "EC:Potential", "EFN", "EFN:Electrons", "EFN:Potential", "EFP", "EFP:Holes", "EFP:Potential", "EG", "EI", "EI:Potential", "EV", "EV:Potential", "ElectronGeneration", "ElectronGeneration:Electrons", "ElectronGeneration:Holes", "Electrons", "HoleGeneration", "HoleGeneration:Electrons", "HoleGeneration:Holes", "Holes", "IntrinsicCharge", "IntrinsicCharge:Potential", "IntrinsicElectrons", "IntrinsicElectrons:Potential", "IntrinsicHoles", "IntrinsicHoles:Potential", "NC", "NCharge", "NCharge:Electrons", "NIE", "NSurfaceNormal_x", "NSurfaceNormal_y", "NTOT", "NV", "NetDoping", "NodeVolume", "OptGen", "PCharge", "PCharge:Holes", "Potential", "PotentialIntrinsicCharge", "PotentialIntrinsicCharge:Potential", "PotentialNodeCharge", "PotentialNodeCharge:Electrons", "PotentialNodeCharge:Holes", "SurfaceArea", "Tn", "USRH_bulk", "USRH_bulk:Electrons", "USRH_bulk:Holes", "USRH_surface", "USRH_surface:Electrons", "USRH_surface:Holes", "V_t", "coordinate_index", "core___via@e_p1nodeelectrons", "core___via@e_p1nodeelectrons:Electrons", "core___via@e_p1nodeholes", "core___via@e_p1nodeholes:Holes", "core___via@e_p1nodemodel", "core___via@e_p1nodemodel:Potential", "ge___via@e_nnodeelectrons", "ge___via@e_nnodeelectrons:Electrons", "ge___via@e_nnodeholes", "ge___via@e_nnodeholes:Holes", "ge___via@e_nnodemodel", "ge___via@e_nnodemodel:Potential", "mu_n_node_lf", "mu_p_node_lf", "node_index", "region_box_at_bulk", "region_box_at_interface_core___box", "region_clad_at_bulk", "region_clad_at_interface_core___clad", "region_clad_at_interface_ge___clad", "region_core_at_bulk", "region_core_at_interface_core___box", "region_core_at_interface_core___clad", "region_core_at_interface_core___ge", "region_ge_at_bulk", "region_ge_at_interface_core___ge", "region_ge_at_interface_ge___clad", "region_via@e_n_at_bulk", "region_via@e_p1_at_bulk", "via@e_n___cladnodemodel", "via@e_n___cladnodemodel:Potential", "via@e_p1___cladnodemodel", "via@e_p1___cladnodemodel:Potential", "DField", "DField:Potential@n0", "DField:Potential@n1", "EField", "EField:Potential@n0", "EField:Potential@n1", "EdgeCouple", "EdgeInverseLength", "EdgeLength", "EdgeNodeVolume", "Electrons@n0", "Electrons@n1", "Epar_n", "Epar_n:Potential@n0", "Epar_n:Potential@n1", "Epar_p", "Epar_p:Potential@n0", "Epar_p:Potential@n1", "Holes@n0", "Holes@n1", "Jn", "Jn:Electrons@n0", "Jn:Electrons@n1", "Jn:Holes@n0", "Jn:Holes@n1", "Jn:Potential@n0", "Jn:Potential@n1", "Jn_arora_lf", "Jn_arora_lf:Electrons@n0", "Jn_arora_lf:Electrons@n1", "Jn_arora_lf:Holes@n0", "Jn_arora_lf:Holes@n1", "Jn_arora_lf:Potential@n0", "Jn_arora_lf:Potential@n1", "Jp", "Jp:Electrons@n0", "Jp:Electrons@n1", "Jp:Holes@n0", "Jp:Holes@n1", "Jp:Potential@n0", "Jp:Potential@n1", "Jp_arora_lf", "Jp_arora_lf:Electrons@n0", "Jp_arora_lf:Electrons@n1", "Jp_arora_lf:Holes@n0", "Jp_arora_lf:Holes@n1", "Jp_arora_lf:Potential@n0", "Jp_arora_lf:Potential@n1", "Potential@n0", "Potential@n1", "V_t_edge", "beta_n", "beta_p", "contactcharge_edge", "contactcharge_edge:Potential@n0", "contactcharge_edge:Potential@n1", "edge_index", "mu_arora_n_lf", "mu_arora_p_lf", "mu_n", "mu_n:Electrons@n0", "mu_n:Electrons@n1", "mu_n:Holes@n0", "mu_n:Holes@n1", "mu_n:Potential@n0", "mu_n:Potential@n1", "mu_p", "mu_p:Electrons@n0", "mu_p:Electrons@n1", "mu_p:Holes@n0", "mu_p:Holes@n1", "mu_p:Potential@n0", "mu_p:Potential@n1", "unitx", "unity", "vsat_n", "vsat_p", "EField_x", "EField_y", "ElementEdgeCouple", "ElementNodeVolume", "Emag", "Jn_x", "Jn_y", "Jnmag", "Jp_x", "Jp_y", "Jpmag"
```

Output filesize: 49.13MB

Command:

`ds.write_devices(file=file, device=device_name, type=type, include="Acceptors|Donors")`

Output file header:

```
TITLE = "450991316286470493"
VARIABLES = "x", "y", "Acceptors", "Donors", "EField_x", "EField_y", "ElementEdgeCouple", "ElementNodeVolume", "Emag", "Jn_x", "Jn_y", "Jnmag", "Jp_x", "Jp_y", "Jpmag"
```

Output filesize: 6.8MB

tcaduser commented:
Would you consider the `include` argument being a list of names or regexes, in case the user wants to just use a python list of names?

Do you think an `exclude` argument for explicitly excluding derivatives and `@n.`, `@en.` models would be useful?

An example of unpacking a string list is `physical_names` in the `create_gmsh_mesh` command.

simbilod commented:
> Would you consider the `include` argument being a list of names or regexes, in case the user wants to just use a python list of names?
> 
> Do you think an `exclude` argument for explicitly excluding derivatives and `@n.`, `@en.` models would be useful?
> 
> An example of unpacking a string list is `physical_names` in the `create_gmsh_mesh` command.

I have implemented both suggestions, and updated the docstring. The default is still to include everything and exclude nothing. If a field is not explicitely included or if a field is explicitely excluded, then it is not recorded.

Here is an example output:

### Default

`ds.write_devices(file=file, device=self.device_name, type=type)`

Output file header:

`VARIABLES = "x", "y", "Acceptors", "AtContactNode", "ContactSurfaceArea", "DEG", "Donors", "EC", "EC:Potential", "EFN", "EFN:Electrons", "EFN:Potential", "EFP", "EFP:Holes", "EFP:Potential", "EG", "EI", "EI:Potential", "EV", "EV:Potential", "ElectronGeneration", "ElectronGeneration:Electrons", "ElectronGeneration:Holes", "Electrons", "HoleGeneration", "HoleGeneration:Electrons", "HoleGeneration:Holes", "Holes", "IntrinsicCharge", "IntrinsicCharge:Potential", "IntrinsicElectrons", "IntrinsicElectrons:Potential", "IntrinsicHoles", "IntrinsicHoles:Potential", "NC", "NCharge", "NCharge:Electrons", "NIE", "NSurfaceNormal_x", "NSurfaceNormal_y", "NTOT", "NV", "NetDoping", "NodeVolume", "OptGen", "PCharge", "PCharge:Holes", "Potential", "PotentialIntrinsicCharge", "PotentialIntrinsicCharge:Potential", "PotentialNodeCharge", "PotentialNodeCharge:Electrons", "PotentialNodeCharge:Holes", "SurfaceArea", "Tn", "USRH_bulk", "USRH_bulk:Electrons", "USRH_bulk:Holes", "USRH_surface", "USRH_surface:Electrons", "USRH_surface:Holes", "V_t", "coordinate_index", "core___via@e_p1nodeelectrons", "core___via@e_p1nodeelectrons:Electrons", "core___via@e_p1nodeholes", "core___via@e_p1nodeholes:Holes", "core___via@e_p1nodemodel", "core___via@e_p1nodemodel:Potential", "ge___via@e_nnodeelectrons", "ge___via@e_nnodeelectrons:Electrons", "ge___via@e_nnodeholes", "ge___via@e_nnodeholes:Holes", "ge___via@e_nnodemodel", "ge___via@e_nnodemodel:Potential", "mu_n_node_lf", "mu_p_node_lf", "node_index", "region_box_at_bulk", "region_box_at_interface_core___box", "region_clad_at_bulk", "region_clad_at_interface_core___clad", "region_clad_at_interface_ge___clad", "region_core_at_bulk", "region_core_at_interface_core___box", "region_core_at_interface_core___clad", "region_core_at_interface_core___ge", "region_ge_at_bulk", "region_ge_at_interface_core___ge", "region_ge_at_interface_ge___clad", "region_via@e_n_at_bulk", "region_via@e_p1_at_bulk", "via@e_n___cladnodemodel", "via@e_n___cladnodemodel:Potential", "via@e_p1___cladnodemodel", "via@e_p1___cladnodemodel:Potential", "DField", "DField:Potential@n0", "DField:Potential@n1", "EField", "EField:Potential@n0", "EField:Potential@n1", "EdgeCouple", "EdgeInverseLength", "EdgeLength", "EdgeNodeVolume", "Electrons@n0", "Electrons@n1", "Epar_n", "Epar_n:Potential@n0", "Epar_n:Potential@n1", "Epar_p", "Epar_p:Potential@n0", "Epar_p:Potential@n1", "Holes@n0", "Holes@n1", "Jn", "Jn:Electrons@n0", "Jn:Electrons@n1", "Jn:Holes@n0", "Jn:Holes@n1", "Jn:Potential@n0", "Jn:Potential@n1", "Jn_arora_lf", "Jn_arora_lf:Electrons@n0", "Jn_arora_lf:Electrons@n1", "Jn_arora_lf:Holes@n0", "Jn_arora_lf:Holes@n1", "Jn_arora_lf:Potential@n0", "Jn_arora_lf:Potential@n1", "Jp", "Jp:Electrons@n0", "Jp:Electrons@n1", "Jp:Holes@n0", "Jp:Holes@n1", "Jp:Potential@n0", "Jp:Potential@n1", "Jp_arora_lf", "Jp_arora_lf:Electrons@n0", "Jp_arora_lf:Electrons@n1", "Jp_arora_lf:Holes@n0", "Jp_arora_lf:Holes@n1", "Jp_arora_lf:Potential@n0", "Jp_arora_lf:Potential@n1", "Potential@n0", "Potential@n1", "V_t_edge", "beta_n", "beta_p", "contactcharge_edge", "contactcharge_edge:Potential@n0", "contactcharge_edge:Potential@n1", "edge_index", "mu_arora_n_lf", "mu_arora_p_lf", "mu_n", "mu_n:Electrons@n0", "mu_n:Electrons@n1", "mu_n:Holes@n0", "mu_n:Holes@n1", "mu_n:Potential@n0", "mu_n:Potential@n1", "mu_p", "mu_p:Electrons@n0", "mu_p:Electrons@n1", "mu_p:Holes@n0", "mu_p:Holes@n1", "mu_p:Potential@n0", "mu_p:Potential@n1", "unitx", "unity", "vsat_n", "vsat_p", "EField_x", "EField_y", "ElementEdgeCouple", "ElementNodeVolume", "Emag", "Jn_x", "Jn_y", "Jnmag", "Jp_x", "Jp_y", "Jpmag"`

Output filesize: 49.13MB

### List of includes only

`ds.write_devices(file=file, device=self.device_name, type=type, include=("Acceptors", ".*Holes.*"))`

Output file header:

`VARIABLES = "x", "y", "Acceptors", "EFP:Holes", "ElectronGeneration:Holes", "HoleGeneration:Holes", "Holes", "IntrinsicHoles", "IntrinsicHoles:Potential", "PCharge:Holes", "PotentialNodeCharge:Holes", "USRH_bulk:Holes", "USRH_surface:Holes", "core___via@e_p1nodeholes:Holes", "ge___via@e_nnodeholes:Holes", "Holes@n0", "Holes@n1", "Jn:Holes@n0", "Jn:Holes@n1", "Jn_arora_lf:Holes@n0", "Jn_arora_lf:Holes@n1", "Jp:Holes@n0", "Jp:Holes@n1", "Jp_arora_lf:Holes@n0", "Jp_arora_lf:Holes@n1", "mu_n:Holes@n0", "mu_n:Holes@n1", "mu_p:Holes@n0", "mu_p:Holes@n1", "EField_x", "EField_y", "ElementEdgeCouple", "ElementNodeVolume", "Emag", "Jn_x", "Jn_y", "Jnmag", "Jp_x", "Jp_y", "Jpmag"`

### List of excludes only

`ds.write_devices(file=file, device=self.device_name, type=type, exclude=(".*@n.*", ".*@en.*"))
`
Output file header:

`VARIABLES = "x", "y", "Acceptors", "AtContactNode", "ContactSurfaceArea", "DEG", "Donors", "EC", "EC:Potential", "EFN", "EFN:Electrons", "EFN:Potential", "EFP", "EFP:Holes", "EFP:Potential", "EG", "EI", "EI:Potential", "EV", "EV:Potential", "ElectronGeneration", "ElectronGeneration:Electrons", "ElectronGeneration:Holes", "Electrons", "HoleGeneration", "HoleGeneration:Electrons", "HoleGeneration:Holes", "Holes", "IntrinsicCharge", "IntrinsicCharge:Potential", "IntrinsicElectrons", "IntrinsicElectrons:Potential", "IntrinsicHoles", "IntrinsicHoles:Potential", "NC", "NCharge", "NCharge:Electrons", "NIE", "NSurfaceNormal_x", "NSurfaceNormal_y", "NTOT", "NV", "NetDoping", "NodeVolume", "OptGen", "PCharge", "PCharge:Holes", "Potential", "PotentialIntrinsicCharge", "PotentialIntrinsicCharge:Potential", "PotentialNodeCharge", "PotentialNodeCharge:Electrons", "PotentialNodeCharge:Holes", "SurfaceArea", "Tn", "USRH_bulk", "USRH_bulk:Electrons", "USRH_bulk:Holes", "USRH_surface", "USRH_surface:Electrons", "USRH_surface:Holes", "V_t", "coordinate_index", "core___via@e_p1nodeelectrons", "core___via@e_p1nodeelectrons:Electrons", "core___via@e_p1nodeholes", "core___via@e_p1nodeholes:Holes", "core___via@e_p1nodemodel", "core___via@e_p1nodemodel:Potential", "ge___via@e_nnodeelectrons", "ge___via@e_nnodeelectrons:Electrons", "ge___via@e_nnodeholes", "ge___via@e_nnodeholes:Holes", "ge___via@e_nnodemodel", "ge___via@e_nnodemodel:Potential", "mu_n_node_lf", "mu_p_node_lf", "node_index", "region_box_at_bulk", "region_box_at_interface_core___box", "region_clad_at_bulk", "region_clad_at_interface_core___clad", "region_clad_at_interface_ge___clad", "region_core_at_bulk", "region_core_at_interface_core___box", "region_core_at_interface_core___clad", "region_core_at_interface_core___ge", "region_ge_at_bulk", "region_ge_at_interface_core___ge", "region_ge_at_interface_ge___clad", "region_via@e_n_at_bulk", "region_via@e_p1_at_bulk", "via@e_n___cladnodemodel", "via@e_n___cladnodemodel:Potential", "via@e_p1___cladnodemodel", "via@e_p1___cladnodemodel:Potential", "DField", "EField", "EdgeCouple", "EdgeInverseLength", "EdgeLength", "EdgeNodeVolume", "Epar_n", "Epar_p", "Jn", "Jn_arora_lf", "Jp", "Jp_arora_lf", "V_t_edge", "beta_n", "beta_p", "contactcharge_edge", "edge_index", "mu_arora_n_lf", "mu_arora_p_lf", "mu_n", "mu_p", "unitx", "unity", "vsat_n", "vsat_p", "EField_x", "EField_y", "ElementEdgeCouple", "ElementNodeVolume", "Emag", "Jn_x", "Jn_y", "Jnmag", "Jp_x", "Jp_y", "Jpmag"`


### Combination of includes and excludes:

`ds.write_devices(file=file, device=self.device_name, type=type, include=(".*Holes.*",), exclude=(".*@n.*", ".*@en.*"))
`

Output file header:

`
VARIABLES = "x", "y", "EFP:Holes", "ElectronGeneration:Holes", "HoleGeneration:Holes", "Holes", "IntrinsicHoles", "IntrinsicHoles:Potential", "PCharge:Holes", "PotentialNodeCharge:Holes", "USRH_bulk:Holes", "USRH_surface:Holes", "core___via@e_p1nodeholes:Holes", "ge___via@e_nnodeholes:Holes", "EField_x", "EField_y", "ElementEdgeCouple", "ElementNodeVolume", "Emag", "Jn_x", "Jn_y", "Jnmag", "Jp_x", "Jp_y", "Jpmag"`

simbilod commented:
I typically keep the commit history as intact as possible, but I am happy to rebase and/or squash if that is a preferred style

tcaduser commented on src/pythonapi/DevsimDoc.cc:
```diff
@@ -1027,7 +1027,7 @@ R"(    devsim.load_devices (file)
 
 static const char write_devices_doc[] =
-R"(    devsim.write_devices (file, device, type)
+R"(    devsim.write_devices (file, device, type, include)
```
The docstrings are actually generated in a different project.
[https://github.com/devsim/devsim_documentation/blob/main/pydoc/meshingCommands.py#L222](https://github.com/devsim/devsim_documentation/blob/main/pydoc/meshingCommands.py#L222)
However I am happy to update this based  on what you have here.

tcaduser commented on src/meshing/TecplotWriter.cc:
```diff
@@ -499,8 +507,17 @@ bool WriteSingleDevice(const std::string &dname, std::ostream &myfile, const std
-            continue;
+          bool is_included = false;
+          is_included = include.empty(); // Assume included if no patterns are specified
+          for (const std::string& pattern : include) {
```
I haven't worked too much with C++ regex.  Is the style similar enough to Python that users will not have any problems?

simbilod replied:
I think it is similar enough, or at least nothing a quick search cannot clarify. Because of the list approach, a user can also simply just list the names of all the fields individually if they wish.

tcaduser commented:
Hi @simbilod,

Thanks for incorporating my suggestions.  This good.  Are you comfortable with the terms with the Apache Software License version 2.  If you are doing this on the behalf of an employer, are they comfortabl with this as well.

simbilod commented:
> Hi @simbilod,
> 
> Thanks for incorporating my suggestions.  This looks good.  Are you comfortable with the terms with the Apache Software License version 2?  If you are doing this on the behalf of an employer, are they comfortable with this as well?

Of course, this contribution is my own. Here is a disclaimer:

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute
this contribution, under the terms of your choice.

tcaduser commented:
> I typically keep the commit history as intact as possible, but I am happy to rebase and/or squash if that is a preferred style

My preference for a commit of this size is to rebase and squash.  There were a bunch of files in the commit history that were accidently added, and would like to avoid them being in the repository history.

For you to squash the commits, I think you would have to do a force push, and I don't have experience how this would affect your pull request.  My understanding from a brief google search is that Github would be able to handle this.

I think I just enabled the ability for you to update your pull request to my incorporation of your `.gitignore` change.  I don't know how that affects your local repo.

I am also able to do a rebase/squash at the end, but I don't know how it will credit your in the repo history.

In the end I am going rebase your PR against the r2.7.4 branch, which is currently identical to main.

simbilod commented:
> > I typically keep the commit history as intact as possible, but I am happy to rebase and/or squash if that is a preferred style
> 
> My preference for a commit of this size is to rebase and squash. There were a bunch of files in the commit history that were accidently added, and would like to avoid them being in the repository history.
> 
> For you to squash the commits, I think you would have to do a force push, and I don't have experience how this would affect your pull request. My understanding from a brief google search is that Github would be able to handle this.
> 
> I think I just enabled the ability for you to update your pull request to my incorporation of your `.gitignore` change. I don't know how that affects your local repo.
> 
> I am also able to do a rebase/squash at the end, but I don't know how it will credit your in the repo history.
> 
> In the end I am going rebase your PR against the r2.7.4 branch, which is currently identical to main.

I squashed everything down to 1 commit, hopefully that is good!

simbilod commented:
> Hi @simbilod,
> 
> Thanks for incorporating my suggestions. This looks good. Are you comfortable with the terms with the Apache Software License version 2? If you are doing this on the behalf of an employer, are they comfortable with this as well?

Also fun fact, according to Github terms of use, [contributions to a repository are assumed to be licensed and licensable under the terms of the repository:](https://docs.github.com/en/site-policy/github-terms/github-terms-of-service#6-contributions-under-repository-license)

tcaduser commented:
> > Hi @simbilod,
> > Thanks for incorporating my suggestions. This looks good. Are you comfortable with the terms with the Apache Software License version 2? If you are doing this on the behalf of an employer, are they comfortable with this as well?
> 
> Also fun fact, according to Github terms of use, [contributions to a repository are assumed to be licensed and licensable under the terms of the repository:](https://docs.github.com/en/site-policy/github-terms/github-terms-of-service#6-contributions-under-repository-license)

Thanks for the tip.  I'm not sure if GitHub is going to be around forever.
